src/python_testing/TC_DeviceBasicComposition.py

```diff
@@ -941,6 +941,217 @@ def log_structured_data(start_tag: str, dump_string):
         log_structured_data('==== json: ', json_str)
         log_structured_data('==== txt: ', txt_str)
 
+    @async_test_body
+    async def test_TC_DESC_2_1(self):
+
+        unique_ids = []
+        ROOT_NODE_DEVICE_TYPE = 0x16
+        server_conformance = 0x001D
+        self.print_step("1a", "TH reads DeviceTypeList and PartsList attributes from DUT for Endpoint 0")
+        parts_list_ep_0 = self.endpoints[0][Clusters.Descriptor][Clusters.Descriptor.Attributes.PartsList]
+        listed_device_types = [i.deviceType for i in self.endpoints[0]
+                               [Clusters.Descriptor][Clusters.Descriptor.Attributes.DeviceTypeList]]
+        device_revision_ep_0 = [i.revision for i in self.endpoints[0]
+                                [Clusters.Descriptor][Clusters.Descriptor.Attributes.DeviceTypeList]]
+        cluster_revision = self.endpoints[0][Clusters.Descriptor][Clusters.Descriptor.Attributes.ClusterRevision]
+        self.print_step("1a.4", "Verify that PartsList attribute is not empty and contains a list of all endpoints.")
+        if len(parts_list_ep_0) == 0:
+            self.fail_current_test()
+        self.print_step("1a.5", "Verify that Revision should match the revision of that Device Type.")
+        if device_revision_ep_0[0] != 1:
```

Comment:
This hard-codes the revision ... should this not be somehow from the spec XMLs so it is dynamic based on spec?